### ISSUE TITLE
Enhance reportPrivateUsage to allow outer class to access nested inner class protected variable

### ISSUE THREAD

**Issue by AnhQuanTrl** [microsoft/pyright#3140]
**Is your feature request related to a problem? Please describe.**
Currently, if outer class attempt to access the protected variable of inner nested class, pyright will report this following `reportPrivateUsage`

**Describe the solution you'd like**
Semantically, an inner nested class should be part of their enclosing class. The outer class still need access to the private/protected variables of the inner to implement its own methods, but want its external users not able to have access the variables in any ways.

**Additional context**
For context, I cannot make `_container` public because I don't want outside code to access it, but since the inner class should be part of the outer class, the outer class should still be able to access inner class' private variables
```python
T = TypeVar("T")
P = TypeVar("P")


class LinkedBinaryTree(BinaryTree[T], Generic[T]):
    class _Node(Generic[P]):
        element: P
        _parent: Optional[LinkedBinaryTree._Node[P]]
        _left: Optional[LinkedBinaryTree._Node[P]]
        _right: Optional[LinkedBinaryTree._Node[P]]

        def __init__(
            self,
            element: P,
            parent: Optional[LinkedBinaryTree._Node[P]] = None,
            left: Optional[LinkedBinaryTree._Node[P]] = None,
            right: Optional[LinkedBinaryTree._Node[P]] = None,
        ) -> None:
            self.element = element
            self._parent = parent
            self._left = left
            self._right = right

    class Position(BinaryTree.Position[P], Generic[P]):
        _node: LinkedBinaryTree._Node[P]
        _container: LinkedBinaryTree[P]

        def __init__(
            self, node: LinkedBinaryTree._Node[P], container: LinkedBinaryTree[P]
        ) -> None:
            self._node = node
            self._container = container

        @property
        def element(self) -> P:
            return self._node.element

        def __eq__(self, __o: object) -> bool:
            return type(__o) is type(self) and __o._node is self._node


    def _validate(self, p: BinaryTree.Position[T]) -> _Node[T]:
        if not isinstance(p, LinkedBinaryTree.Position):
            raise TypeError("p must be proper Position type")
        if p._container is not self:
            # ERROR REPORTED HERE

```

**Comment by erictraut** [comment]
> Semantically, an inner nested class should be part of their enclosing class.

I don't think I agree. A class is a form of encapsulation. If you're using a class and it has variables marked private, code outside of that class should not access them. 

By your logic, if a module declares a class, that class is "semantically part of the module", so code at the module level should be allowed to access private variables within that class.

The "reportPrivateUsage" diagnostic check is off by default. I presume you have enabled it explicitly or enabled "strict" type checking. You have the option of selectively disabling this check on a per-file basis using a comment (`# pyright: reportPrivateUsage = false`) if you don't want it to apply to particular pieces of code.

**Comment by debonte** [comment]
It's interesting to note that a nested class has access to private variables on its parent class, but the reverse is not true.

For what it's worth, this matches the behavior of C#. I'm just using C# as an example because I'm familiar with it. I'm guessing this is true in other languages as well.
